File: quanta-view-backend/src/candleAggregator.ts
import { FinnhubTrade, Candlestick } from './types/socket';
import { Server } from 'socket.io';

// Stores the currently forming candle for each symbol
const currentCandles: { [symbol: string]: Candlestick | undefined } = {};

// Stores a history of completed candles for each symbol
const stockCandleHistory: { [symbol: string]: Candlestick[] } = {};

const CANDLE_INTERVAL_SECONDS = 60; 
const MAX_CANDLE_HISTORY = 500;

/**
 * Normalizes a timestamp to the start of the current candle interval.
 * @param timestampMs The original timestamp in seconds.
 * @param intervalSeconds The candle interval in seconds.
 * @returns The normalized timestamp (start of the candle interval) in seconds.
 */
function normalizeTimestamp(timestamps: number, intervalSeconds: number): number {
  return Math.floor(timestamps / (intervalSeconds)) * (intervalSeconds);
}

/**
 * Processes a new Finnhub trade and updates/completes candlesticks.
 * @param trade The incoming Finnhub trade object.
 * @param io Socket.IO server instance to emit completed candles.
 */
export function processNewTrade(trade: FinnhubTrade, io: Server) {
  const symbol = trade.s;
  const price = trade.p;
  const volume = trade.v;

  // Finnhub timestamps are in seconds, convert to milliseconds
  const tradeTimestampMs = trade.t;
  const currentIntervalStart = normalizeTimestamp(tradeTimestampMs, CANDLE_INTERVAL_SECONDS);

  let currentCandle = currentCandles[symbol];

  // 1. Initialize new candle if not exists or if it's a new interval
  if (!currentCandle || currentCandle.time !== currentIntervalStart) {
    // If there was an old candle, it's now complete (unless it was just initialized)
    if (currentCandle && currentCandle.time < currentIntervalStart) {
      // Old candle is complete. Store it and emit.
      if (!stockCandleHistory[symbol]) {
        stockCandleHistory[symbol] = [];
      }
      stockCandleHistory[symbol].push(currentCandle);

      // Keep history limited
      if (stockCandleHistory[symbol].length > MAX_CANDLE_HISTORY) {
        stockCandleHistory[symbol].shift(); // Remove the oldest
      }

      // Emit the completed candle to all connected clients
      io.emit('candle-completed', currentCandle);
      console.log(`[Candle Aggregator] Emitted completed ${CANDLE_INTERVAL_SECONDS/60}-min candle for ${symbol}:`, currentCandle);
    }

    // Start a new candle for the current interval
    currentCandles[symbol] = {
      time: currentIntervalStart,
      open: price,
      high: price,
      low: price,
      close: price,
      volume: volume || 0 // Ensure volume is a number
    };
    console.log(`[Candle Aggregator] Started new ${CANDLE_INTERVAL_SECONDS/60}-min candle for ${symbol} at ${new Date(currentIntervalStart).toLocaleTimeString()}`);

  } else {
    // 2. Update existing candle
    currentCandle.high = Math.max(currentCandle.high, price);
    currentCandle.low = Math.min(currentCandle.low, price);
    currentCandle.close = price;
    currentCandle.volume += (volume || 0);
  }
}

/**
 * Returns the current in-memory history of completed candles for a symbol.
 * @param symbol The stock symbol.
 * @returns An array of Candlestick objects.
 */
export function getCandleHistory(symbol: string): Candlestick[] {
  return stockCandleHistory[symbol] || [];
}

/**
 * Allows external access to the currently forming candles for debugging or live updates.
 * @returns A map of currently forming candles.
 */
export function getCurrentFormingCandles(): { [symbol: string]: Candlestick | undefined } {
  return currentCandles;
}

File: quanta-view-backend/candle_aggregator.py
# candle_aggregator.py (Corrected version)
from datetime import datetime
from collections import defaultdict
from typing import Dict, List, Any, Callable, Awaitable

# Candle state
candle_history: Dict[str, List[Dict[str, Any]]] = defaultdict(list)
current_forming: Dict[str, Dict[str, Any]] = {}

CANDLE_INTERVAL = 60  # seconds (1-minute candle)

def floor_timestamp_to_interval(ts: int) -> int:
    return ts - (ts % (CANDLE_INTERVAL * 1000))

async def process_new_trade(
    trade: Dict[str, Any], 
    on_forming_candle: Callable,
    on_completed_candle: Callable
):
    symbol = trade['s']
    price = trade['p']
    volume = trade['v']
    timestamp = trade['t']

    bucket_ts = floor_timestamp_to_interval(timestamp)

    candle = current_forming.get(symbol)
    
    if not candle or candle['timestamp'] != bucket_ts:
        # A new candle is starting. The old one is now complete.
        if candle:
            candle_history[symbol].append(candle)
            await on_completed_candle(symbol, candle) # <-- Now correctly awaited
            
        # Create the new forming candle
        candle = {
            'timestamp': bucket_ts,
            'open': price,
            'high': price,
            'low': price,
            'close': price,
            'volume': volume
        }
        current_forming[symbol] = candle
    else:
        # We are still in the same candle interval
        candle['high'] = max(candle['high'], price)
        candle['low'] = min(candle['low'], price)
        candle['close'] = price
        candle['volume'] += volume

    # Send the updated forming candle to the frontend
    await on_forming_candle(symbol, candle) # <-- Now correctly awaited

def get_candle_history(symbol: str) -> List[Dict[str, Any]]:
    history = sorted(candle_history[symbol], key= lambda x: x['timestamp'])
    return history

def get_current_forming_candle(symbol: str) -> Dict[str, Any]:
    return current_forming.get(symbol)